crates/ruff_cli/src/commands/format.rs

```diff
@@ -78,9 +110,80 @@ fn format_path(path: &Path, options: PyFormatOptions) -> Result<(), FormatterIte
         format_module(&unformatted, options)
             .map_err(|err| FormatterIterationError::FormatModule(path.to_path_buf(), err))?
     };
-    std::fs::write(path, formatted.as_code().as_bytes())
-        .map_err(|err| FormatterIterationError::Write(path.to_path_buf(), err))?;
-    Ok(())
+    let formatted = formatted.as_code();
+    if formatted.len() == unformatted.len() && formatted == unformatted {
+        Ok(FormatResult::Unchanged)
+    } else {
+        std::fs::write(path, formatted.as_bytes())
+            .map_err(|err| FormatterIterationError::Write(path.to_path_buf(), err))?;
+        Ok(FormatResult::Formatted)
+    }
+}
+
+#[derive(Debug, Clone, Copy)]
+enum FormatResult {
+    /// The file was formatted.
+    Formatted,
+    /// The file was unchanged, as the formatted contents matched the existing contents.
+    Unchanged,
+    /// The file was skipped, as it was not a Python file.
+    Skipped,
+}
+
+#[derive(Debug, Default)]
+struct FormatResultSummary {
+    /// The number of files that were formatted.
+    formatted: usize,
+    /// The number of files that were unchanged.
+    unchanged: usize,
+    /// The number of files that were skipped.
+    skipped: usize,
+}
+
+impl From<Vec<FormatResult>> for FormatResultSummary {
+    fn from(diagnostics: Vec<FormatResult>) -> Self {
+        let mut path_diagnostics = Self::default();
+        for diagnostic in diagnostics {
+            match diagnostic {
+                FormatResult::Formatted => path_diagnostics.formatted += 1,
+                FormatResult::Unchanged => path_diagnostics.unchanged += 1,
+                FormatResult::Skipped => path_diagnostics.skipped += 1,
+            }
+        }
+        path_diagnostics
+    }
+}
+
+impl FormatResultSummary {
+    /// Pretty-print a [`FormatResultSummary`] for user-facing display.
+    fn show_user(&self, writer: &mut dyn Write) -> Result<(), io::Error> {
+        if self.formatted > 0 && self.unchanged > 0 {
+            writeln!(
+                writer,
+                "{} file{} reformatted, {} file{} left unchanged",
```

Comment:
Is it clear that these files are not ones that were excluded for some reason? Should we say "already formatted"?

Reply:
I erred on the side of matching Black's terminology here.

Reply:
Alright. Too be considered another day :)